storage/client/internal/parse_rfc3339.cc

```diff
@@ -0,0 +1,181 @@
+                " the base date and time portion.");
+  }
+  if (date_time_separator != 'T' and date_time_separator != 't') {
+    ReportError(timestamp, "Invalid date-time separator, expected 'T' or 't'.");
+  }
+  if (mon <= 0 or mon > 12) {
+    ReportError(timestamp, "Out of range month.");
+  }
+  if (mday <= 0 or mday > 31) {
+    ReportError(timestamp, "Out of range month day.");
+  }
+  if (hh < 0 or hh > 23) {
+    ReportError(timestamp, "Out of range hour.");
+  }
+  if (mm < 0 or mm > 59) {
+    ReportError(timestamp, "Out of range minute.");
+  }
+  if (ss < 0 or ss > 60) {
+    ReportError(timestamp, "Out of range second.");
+  }
+  // Advance the pointer for all the characters read.
+  buffer += pos;
+
+  std::tm tm{0};
+  tm.tm_year = year - 1900;
+  tm.tm_mon = mon - 1;
+  tm.tm_mday = mday;
+  tm.tm_hour = hh;
+  tm.tm_min = mm;
+  tm.tm_sec = ss;
+  return std::chrono::system_clock::from_time_t(std::mktime(&tm));
+}
+
+std::chrono::nanoseconds ParseFractionalSeconds(char const*& buffer,
+                                                std::string const& timestamp) {
+  if (buffer[0] != '.') {
+    return std::chrono::nanoseconds(0);
+  }
+  ++buffer;
+
+  long fractional_seconds;
+  int pos;
+  auto count = std::sscanf(buffer, "%9ld%n", &fractional_seconds, &pos);
+  if (count != 1) {
+    ReportError(timestamp, "Invalid fractional seconds component.");
+  }
+  // Normalize the fractional seconds to nanoseconds.
+  for (int digits = pos; digits < 9; ++digits) {
+    fractional_seconds *= 10;
+  }
+  // Now skip any other digits ...
+  buffer += pos;
+  while (std::isdigit(buffer[0])) {
+    ++buffer;
+  }
+  return std::chrono::nanoseconds(fractional_seconds);
+}
+
+std::chrono::seconds ParseOffset(char const*& buffer,
+                                 std::string const& timestamp) {
+  if (buffer[0] == '+' or buffer[0] == '-') {
+    bool positive = buffer[0] == '+';
+    ++buffer;
+    // Parse the HH:MM offset.
+    int hh, mm, pos;
+    auto count = std::sscanf(buffer, "%2d:%2d%n", &hh, &mm, &pos);
+    constexpr int EXPECTED_OFFSET_WIDTH = 5;
+    constexpr int EXPECTED_OFFSET_FIELDS = 2;
+    if (count != EXPECTED_OFFSET_FIELDS or pos != EXPECTED_OFFSET_WIDTH) {
+      ReportError(timestamp, "Invalid timezone offset, expected [+/-]HH:SS.");
+    }
+    if (hh < 0 or hh > 23) {
+      ReportError(timestamp, "Out of range offset hour.");
+    }
+    if (mm < 0 or mm > 59) {
+      ReportError(timestamp, "Out of range offset minute.");
+    }
+    buffer += pos;
+    using namespace std::chrono;
+    if (positive) {
+      return duration_cast<seconds>(hours(hh) + minutes(mm));
+    }
+    return duration_cast<seconds>(-hours(hh) - minutes(mm));
+  }
+  if (buffer[0] != 'Z' and buffer[0] != 'z') {
+    ReportError(timestamp, "Invalid timezone offset, expected 'Z' or 'z'.");
+  }
+  ++buffer;
+  return std::chrono::seconds(0);
+}
+}  // anonymous namespace
+
+namespace storage {
```

Comment:
Why is this part of the `storage` directory and namespace? Isn't this generic, and should be in an internal directory common to all clients? Or is it here because only `storage` needs it, and you plan to refactor it as soon as you have a second user?

Reply:
Correct, I would refactor on the second use.